File: backend/agent/tools/screen.py
from __future__ import annotations

import asyncio
import datetime
import os
import subprocess
import time
from pathlib import Path
from typing import Any, Dict

from agent.tools.base import Tool
from agent.tools.input import (
	_get_screenshot_dir,
	_cleanup_old_screenshots,
	_map_to_model,
	_map_to_screen,
)


def _is_windows() -> bool:
	return os.name == "nt"


class ImageLocateTool(Tool):
	name = "image_locate"
	description = (
		"Find a template image on the screen and return the center coordinates "
		"in screenshot space (ready to pass to the mouse tool). "
		"Use to click UI elements reliably instead of guessing pixels. "
		"Returns 'not found' if the template is not visible."
	)
	params = {
		"image": {"type": "string", "description": "Path to the template image (PNG/JPG)"},
		"confidence": {
			"type": "number",
			"description": "Match tolerance 0..1 (default 0.9, needs opencv; falls back to exact match)",
		},
	}

	def tool_definition(self) -> dict:
		d = super().tool_definition()
		d["function"]["parameters"]["required"] = ["image"]
		return d

	async def execute(self, args: Dict[str, Any]) -> str:
		path = str(args.get("image", "")).strip()
		if not path:
			return "error: image path is required"
		if not Path(path).exists():
			return f"error: image not found: {path}"
		confidence = args.get("confidence")
		if confidence is not None:
			try:
				confidence = float(confidence)
			except (ValueError, TypeError):
				return "error: confidence must be a number between 0 and 1"
			if not (0 <= confidence <= 1):
				return "error: confidence must be between 0 and 1"
		try:
			import pyautogui

			def _locate():
				try:
					if confidence is not None:
						return pyautogui.locateOnScreen(path, confidence=confidence)
					return pyautogui.locateOnScreen(path)
				except TypeError:
					# older pyautogui without confidence / opencv missing
					return pyautogui.locateOnScreen(path)

			box = await asyncio.to_thread(_locate)
		except ImportError:
			return "error: pip install pyautogui (opencv-python for confidence)"
		except Exception as e:
			# pyautogui raises ImageNotFoundException (subclass of Exception) when missing
			if "ImageNotFound" in type(e).__name__ or "could not be located" in str(e).lower():
				return "not found"
			return f"error: {e}"

		if box is None:
			return "not found"
		rx, ry = int(box.left + box.width / 2), int(box.top + box.height / 2)
		mx, my = _map_to_model(rx, ry)
		return f"found at ({mx}, {my}) — size {int(box.width)}x{int(box.height)}, screen ({rx}, {ry})"


# ── OCR via Windows.Media.Ocr (no extra pip deps; uses the OS engine) ────────

_OCR_SCRIPT = r"""
$Path = $env:WARDEN_OCR_PATH
Add-Type -AssemblyName System.Runtime.WindowsRuntime | Out-Null
$asTaskGeneric = ([System.WindowsRuntimeSystemExtensions].GetMethods() | Where-Object {
	$_.Name -eq 'AsTask' -and $_.GetParameters().Count -eq 1 -and
	$_.GetParameters()[0].ParameterType.Name -eq 'IAsyncOperation`1'
})[0]
function Await($op, $resultType) {
	$asTask = $asTaskGeneric.MakeGenericMethod($resultType)
	$task = $asTask.Invoke($null, @($op))
	$task.Wait(-1) | Out-Null
	$task.Result
}
[Windows.Media.Ocr.OcrEngine, Windows.Foundation, ContentType = WindowsRuntime] | Out-Null
[Windows.Graphics.Imaging.BitmapDecoder, Windows.Foundation, ContentType = WindowsRuntime] | Out-Null
[Windows.Storage.StorageFile, Windows.Foundation, ContentType = WindowsRuntime] | Out-Null
$engine = [Windows.Media.Ocr.OcrEngine]::TryCreateFromUserProfileLanguages()
if ($null -eq $engine) { Write-Error 'OCR engine unavailable (no language pack installed)'; exit 1 }
$file = Await ([Windows.Storage.StorageFile]::GetFileFromPathAsync($Path)) ([Windows.Storage.StorageFile])
$stream = Await ($file.OpenAsync([Windows.Storage.FileAccessMode]::Read)) ([Windows.Storage.Streams.IRandomAccessStream])
$decoder = Await ([Windows.Graphics.Imaging.BitmapDecoder]::CreateAsync($stream)) ([Windows.Graphics.Imaging.BitmapDecoder])
$bitmap = Await ($decoder.GetSoftwareBitmapAsync()) ([Windows.Graphics.Imaging.SoftwareBitmap])
$result = Await ($engine.RecognizeAsync($bitmap)) ([Windows.Media.Ocr.OcrResult])
[Console]::OutputEncoding = [System.Text.Encoding]::UTF8
Write-Output $result.Text
"""


async def _ocr_image(path: str) -> str:
	"""Run Windows OCR on an image file, returning recognized text."""
	env = dict(os.environ)
	env["WARDEN_OCR_PATH"] = path
	proc = await asyncio.to_thread(
		subprocess.run,
		["powershell", "-NoProfile", "-NonInteractive", "-Command", _OCR_SCRIPT],
		capture_output=True, text=True, timeout=30,
		encoding="utf-8", errors="replace", env=env,
	)
	if proc.returncode != 0:
		raise RuntimeError((proc.stderr or "OCR failed").strip().splitlines()[0] if proc.stderr else "OCR failed")
	return (proc.stdout or "").strip()


def _capture_region(region: tuple[int, int, int, int] | None) -> str:
	"""Grab the screen (or a region) to a temp PNG and return its path."""
	from PIL import ImageGrab
	screenshot_dir = _get_screenshot_dir()
	_cleanup_old_screenshots(screenshot_dir, max_age_seconds=300)
	name = screenshot_dir / f"ocr_{datetime.datetime.now():%Y%m%d_%H%M%S_%f}.png"
	img = ImageGrab.grab()
	if region is not None:
		x, y, w, h = region
		img = img.crop((x, y, x + w, y + h))
	img.save(name)
	return str(name)


class OcrTool(Tool):
	name = "ocr"
	description = (
		"Recognize text on the screen or in an image file using the Windows OCR "
		"engine (no extra dependencies). Windows only. "
		"Give an image path, or omit it to OCR the current screen. "
		"Optional x/y/w/h restricts OCR to a region (screenshot space)."
	)
	params = {
		"image": {"type": "string", "description": "Path to an image to OCR (optional; defaults to a screen grab)"},
		"x": {"type": "integer", "description": "Region left (screenshot space, optional)"},
		"y": {"type": "integer", "description": "Region top (screenshot space, optional)"},
		"w": {"type": "integer", "description": "Region width (optional)"},
		"h": {"type": "integer", "description": "Region height (optional)"},
	}

	def tool_definition(self) -> dict:
		d = super().tool_definition()
		d["function"]["parameters"]["required"] = []
		return d

	async def execute(self, args: Dict[str, Any]) -> str:
		if not _is_windows():
			return "error: ocr is Windows-only"
		path = str(args.get("image", "")).strip()
		try:
			if not path:
				region = None
				if all(k in args for k in ("x", "y", "w", "h")):
					rx, ry = _map_to_screen(int(args["x"]), int(args["y"]))
					rw, rh = _map_to_screen(int(args["w"]), int(args["h"]))
					region = (rx, ry, rw, rh)
				path = await asyncio.to_thread(_capture_region, region)
			elif not Path(path).exists():
				return f"error: image not found: {path}"
			text = await _ocr_image(path)
		except ImportError:
			return "error: pip install Pillow"
		except subprocess.TimeoutExpired:
			return "error: timeout running OCR"
		except Exception as e:
			return f"error: {e}"
		return text or "(no text recognized)"


class WaitForTool(Tool):
	name = "wait_for"
	description = (
		"Wait until a condition holds, polling until it's true or the timeout "
		"elapses. Prefer this over fixed sleeps for reliable automation. "
		"type: window (title appears), text (substring visible on screen, needs OCR), "
		"image (template found on screen)."
	)
	params = {
		"type": {"type": "string", "description": "window | text | image"},
		"target": {"type": "string", "description": "Window title substring, text to find, or template image path"},
		"timeout": {"type": "number", "description": "Max seconds to wait (default 10, max 30)"},
		"interval": {"type": "number", "description": "Seconds between checks (default 0.5)"},
	}

	def tool_definition(self) -> dict:
		d = super().tool_definition()
		d["function"]["parameters"]["required"] = ["type", "target"]
		return d

	async def execute(self, args: Dict[str, Any]) -> str:
		kind = str(args.get("type", "")).strip().lower()
		target = str(args.get("target", "")).strip()
		if not target:
			return "error: target is required"
		timeout = min(float(args.get("timeout", 10)), 30.0)
		interval = max(float(args.get("interval", 0.5)), 0.1)
		if kind not in ("window", "text", "image"):
			return "error: type must be window, text, or image"

		start = time.monotonic()
		while True:
			try:
				hit = await self._check(kind, target)
			except Exception as e:
				return f"error: {e}"
			elapsed = time.monotonic() - start
			if hit:
				return f"found after {elapsed:.1f}s"
			if elapsed + interval >= timeout:
				return f"timeout: '{target}' not found after {timeout:.0f}s"
			await asyncio.sleep(interval)

	async def _check(self, kind: str, target: str) -> bool:
		if kind == "window":
			from agent.tools.window import _enumerate_windows
			needle = target.lower()
			return any(needle in str(w.get("title", "")).lower() for w in await _enumerate_windows())
		if kind == "image":
			import pyautogui

			def _locate():
				try:
					return pyautogui.locateOnScreen(target)
				except Exception as e:
					if "ImageNotFound" in type(e).__name__ or "could not be located" in str(e).lower():
						return None
					raise

			return (await asyncio.to_thread(_locate)) is not None
		# text
		path = await asyncio.to_thread(_capture_region, None)
		text = await _ocr_image(path)
		return target.lower() in text.lower()

File: agent/tools/screen.go
package tools

import (
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
	"time"
)

// ── OCR PowerShell script (Windows.Media.Ocr) ───────────────────────────────
// Identical to the Python version — no extra deps, uses the OS OCR engine.

const ocrScript = `
$Path = $env:WARDEN_OCR_PATH
Add-Type -AssemblyName System.Runtime.WindowsRuntime | Out-Null
$asTaskGeneric = ([System.WindowsRuntimeSystemExtensions].GetMethods() | Where-Object {
	$_.Name -eq 'AsTask' -and $_.GetParameters().Count -eq 1 -and
	$_.GetParameters()[0].ParameterType.Name -eq 'IAsyncOperation` + "`" + `1'
})[0]
function Await($op, $resultType) {
	$asTask = $asTaskGeneric.MakeGenericMethod($resultType)
	$task = $asTask.Invoke($null, @($op))
	$task.Wait(-1) | Out-Null
	$task.Result
}
[Windows.Media.Ocr.OcrEngine, Windows.Foundation, ContentType = WindowsRuntime] | Out-Null
[Windows.Graphics.Imaging.BitmapDecoder, Windows.Foundation, ContentType = WindowsRuntime] | Out-Null
[Windows.Storage.StorageFile, Windows.Foundation, ContentType = WindowsRuntime] | Out-Null
$engine = [Windows.Media.Ocr.OcrEngine]::TryCreateFromUserProfileLanguages()
if ($null -eq $engine) { Write-Error 'OCR engine unavailable (no language pack installed)'; exit 1 }
$file = Await ([Windows.Storage.StorageFile]::GetFileFromPathAsync($Path)) ([Windows.Storage.StorageFile])
$stream = Await ($file.OpenAsync([Windows.Storage.FileAccessMode]::Read)) ([Windows.Storage.Streams.IRandomAccessStream])
$decoder = Await ([Windows.Graphics.Imaging.BitmapDecoder]::CreateAsync($stream)) ([Windows.Graphics.Imaging.BitmapDecoder])
$bitmap = Await ($decoder.GetSoftwareBitmapAsync()) ([Windows.Graphics.Imaging.SoftwareBitmap])
$result = Await ($engine.RecognizeAsync($bitmap)) ([Windows.Media.Ocr.OcrResult])
[Console]::OutputEncoding = [System.Text.Encoding]::UTF8
Write-Output $result.Text
`

func runOCR(imagePath string) (string, error) {
	shell := shellExe()
	cmd := exec.Command(shell, "-NoProfile", "-NonInteractive", "-Command", ocrScript)
	cmd.Env = append(os.Environ(), "WARDEN_OCR_PATH="+imagePath)
	out, err := cmd.Output()
	if err != nil {
		if ee, ok := err.(*exec.ExitError); ok {
			stderr := strings.TrimSpace(string(ee.Stderr))
			if stderr != "" {
				lines := strings.SplitN(stderr, "\n", 2)
				return "", fmt.Errorf("%s", lines[0])
			}
		}
		return "", fmt.Errorf("OCR failed: %v", err)
	}
	return strings.TrimSpace(string(out)), nil
}

// captureRegionToFile takes a screenshot (optionally cropped) and saves to temp.
// region: [x, y, w, h] in real screen pixels, or nil for full screen.
func captureRegionToFile(region []int) (string, error) {
	dir := screenshotDir()
	cleanupOldScreenshots(dir, 5*time.Minute)

	name := filepath.Join(dir, fmt.Sprintf("ocr_%s.png", time.Now().Format("20060102_150405_000")))

	if region != nil && len(region) == 4 {
		// Use PowerShell to capture and crop — avoids importing image processing in Go
		script := fmt.Sprintf(`
Add-Type -AssemblyName System.Drawing | Out-Null
Add-Type -AssemblyName System.Windows.Forms | Out-Null
$bounds = [System.Windows.Forms.Screen]::PrimaryScreen.Bounds
$bmp = New-Object System.Drawing.Bitmap($bounds.Width, $bounds.Height)
$g = [System.Drawing.Graphics]::FromImage($bmp)
$g.CopyFromScreen($bounds.Location, [System.Drawing.Point]::Empty, $bounds.Size)
$g.Dispose()
$crop = $bmp.Clone((New-Object System.Drawing.Rectangle(%d,%d,%d,%d)), $bmp.PixelFormat)
$bmp.Dispose()
$crop.Save('%s', [System.Drawing.Imaging.ImageFormat]::Png)
$crop.Dispose()
`, region[0], region[1], region[2], region[3], strings.ReplaceAll(name, "'", "''"))
		shell := shellExe()
		cmd := exec.Command(shell, "-NoProfile", "-NonInteractive", "-Command", script)
		if err := cmd.Run(); err != nil {
			return "", fmt.Errorf("capture failed: %v", err)
		}
	} else {
		// Full screen capture via PowerShell GDI+
		script := fmt.Sprintf(`
Add-Type -AssemblyName System.Drawing | Out-Null
Add-Type -AssemblyName System.Windows.Forms | Out-Null
$bounds = [System.Windows.Forms.Screen]::PrimaryScreen.Bounds
$bmp = New-Object System.Drawing.Bitmap($bounds.Width, $bounds.Height)
$g = [System.Drawing.Graphics]::FromImage($bmp)
$g.CopyFromScreen($bounds.Location, [System.Drawing.Point]::Empty, $bounds.Size)
$g.Dispose()
$bmp.Save('%s', [System.Drawing.Imaging.ImageFormat]::Png)
$bmp.Dispose()
`, strings.ReplaceAll(name, "'", "''"))
		shell := shellExe()
		cmd := exec.Command(shell, "-NoProfile", "-NonInteractive", "-Command", script)
		if err := cmd.Run(); err != nil {
			return "", fmt.Errorf("capture failed: %v", err)
		}
	}
	return name, nil
}

// ── ImageLocateTool ──────────────────────────────────────────────────────────

// PowerShell GDI+ template matching script. Scans pixel-by-pixel.
const imageLocateScript = `
param($TemplatePath, $ScreenshotPath)
Add-Type -AssemblyName System.Drawing | Out-Null
$tmpl = [System.Drawing.Bitmap]::new($TemplatePath)
$screen = [System.Drawing.Bitmap]::new($ScreenshotPath)
$tw = $tmpl.Width; $th = $tmpl.Height
$sw = $screen.Width; $sh = $screen.Height
$found = $false
for ($sy = 0; $sy -le ($sh - $th); $sy++) {
	for ($sx = 0; $sx -le ($sw - $tw); $sx++) {
		$match = $true
		for ($ty = 0; $ty -lt $th -and $match; $ty += [Math]::Max(1, [int]($th/8))) {
			for ($tx = 0; $tx -lt $tw -and $match; $tx += [Math]::Max(1, [int]($tw/8))) {
				$sp = $screen.GetPixel($sx+$tx, $sy+$ty)
				$tp = $tmpl.GetPixel($tx, $ty)
				if ([Math]::Abs([int]$sp.R-[int]$tp.R) -gt 30 -or
					[Math]::Abs([int]$sp.G-[int]$tp.G) -gt 30 -or
					[Math]::Abs([int]$sp.B-[int]$tp.B) -gt 30) {
					$match = $false
				}
			}
		}
		if ($match) {
			# Verify with full pixel check on a sample
			$cx = $sx + [int]($tw/2); $cy = $sy + [int]($th/2)
			Write-Output "$cx,$cy,$tw,$th,$sx,$sy"
			$found = $true; break
		}
	}
	if ($found) { break }
}
if (-not $found) { Write-Output "NOTFOUND" }
$tmpl.Dispose(); $screen.Dispose()
`

type ImageLocateTool struct{}

func (t *ImageLocateTool) Name() string { return "image_locate" }

func (t *ImageLocateTool) Spec() ToolSpec {
	return ToolSpec{
		Description: "Locate a sub-image on screen and return its coordinates.",
		Params: map[string]any{
			"image": prop("string", "Absolute path to the template image to find on screen"),
		},
		Required: []string{"image"},
	}
}

func (t *ImageLocateTool) Execute(args map[string]any) Result {
	path := strings.TrimSpace(getStr(args, "image"))
	if path == "" {
		return R("error: image path is required")
	}
	if _, err := os.Stat(path); os.IsNotExist(err) {
		return R("error: image not found: " + path)
	}

	// Take a fresh screenshot for comparison
	screenshotPath, err := captureRegionToFile(nil)
	if err != nil {
		return R("error: " + err.Error())
	}
	defer os.Remove(screenshotPath)

	shell := shellExe()
	cmd := exec.Command(shell, "-NoProfile", "-NonInteractive", "-Command", imageLocateScript,
		"-TemplatePath", path, "-ScreenshotPath", screenshotPath)
	out, err := cmd.Output()
	if err != nil {
		return R("error: image locate failed: " + err.Error())
	}

	result := strings.TrimSpace(string(out))
	if result == "NOTFOUND" || result == "" {
		return R("not found")
	}

	// Parse "cx,cy,tw,th,sx,sy"
	var cx, cy, tw, th, sx, sy int
	n, _ := fmt.Sscanf(result, "%d,%d,%d,%d,%d,%d", &cx, &cy, &tw, &th, &sx, &sy)
	if n < 4 {
		return R("not found")
	}
	mx, my := mapToModel(cx, cy)
	return R(fmt.Sprintf("found at (%d, %d) — size %dx%d, screen (%d, %d)", mx, my, tw, th, cx, cy))
}

// ── OcrTool ──────────────────────────────────────────────────────────────────

type OcrTool struct{}

func (t *OcrTool) Name() string { return "ocr" }

func (t *OcrTool) Spec() ToolSpec {
	return ToolSpec{
		Description: "Recognize text on screen or in an image file.",
		Params: map[string]any{
			"image": prop("string", "Path to image file; omit to capture the screen"),
			"x":     prop("integer", "Region X (screenshot space, requires y/w/h)"),
			"y":     prop("integer", "Region Y"),
			"w":     prop("integer", "Region width"),
			"h":     prop("integer", "Region height"),
		},
	}
}

func (t *OcrTool) Execute(args map[string]any) Result {
	path := strings.TrimSpace(getStr(args, "image"))

	if path == "" {
		// Capture the screen, optionally a region
		var region []int
		if _, hasX := args["x"]; hasX {
			if _, hasY := args["y"]; hasY {
				if _, hasW := args["w"]; hasW {
					if _, hasH := args["h"]; hasH {
						rx, ry := mapToScreen(getInt(args, "x", 0), getInt(args, "y", 0))
						rw, rh := mapToScreen(getInt(args, "w", 0), getInt(args, "h", 0))
						region = []int{rx, ry, rw, rh}
					}
				}
			}
		}
		var err error
		path, err = captureRegionToFile(region)
		if err != nil {
			return R("error: " + err.Error())
		}
		defer os.Remove(path)
	} else {
		if _, err := os.Stat(path); os.IsNotExist(err) {
			return R("error: image not found: " + path)
		}
	}

	text, err := runOCR(path)
	if err != nil {
		return R("error: " + err.Error())
	}
	if text == "" {
		return R("(no text recognized)")
	}
	return R(text)
}

// --- ScreenshotRegionTool ---

type ScreenshotRegionTool struct{}

func (t *ScreenshotRegionTool) Name() string { return "screenshot_region" }

func (t *ScreenshotRegionTool) Spec() ToolSpec {
	return ToolSpec{
		Description: "Capture a rectangular screen region to a PNG. Coordinates are in screenshot space.",
		Params: map[string]any{
			"x": prop("integer", "Region top-left X (screenshot space)"),
			"y": prop("integer", "Region top-left Y (screenshot space)"),
			"w": prop("integer", "Region width (screenshot space)"),
			"h": prop("integer", "Region height (screenshot space)"),
		},
		Required: []string{"x", "y", "w", "h"},
	}
}

func (t *ScreenshotRegionTool) Execute(args map[string]any) Result {
	w := getInt(args, "w", 0)
	h := getInt(args, "h", 0)
	if w <= 0 || h <= 0 {
		return R("error: w and h must be positive")
	}
	rx, ry := mapToScreen(getInt(args, "x", 0), getInt(args, "y", 0))
	sw, sh := screenSize()
	scale := scaleFactor(sw, sh)
	rw, rh := w, h
	if scale < 1.0 {
		rw = int(float64(w) / scale)
		rh = int(float64(h) / scale)
	}
	path, err := captureRegionToFile([]int{rx, ry, rw, rh})
	if err != nil {
		return R("error: " + err.Error())
	}
	return R("saved: " + path)
}

// --- WaitForTool ---

type WaitForTool struct{}

func (t *WaitForTool) Name() string { return "wait_for" }

func (t *WaitForTool) Spec() ToolSpec {
	return ToolSpec{
		Description: "Poll until a UI condition (window, text, or image) is met.",
		Params: map[string]any{
			"type":     prop("string", "window, text, or image"),
			"target":   prop("string", "Window title, text to find, or image path"),
			"timeout":  prop("number", "Seconds to wait (default 10, max 30)"),
			"interval": prop("number", "Poll interval in seconds (default 0.5)"),
		},
		Required: []string{"type", "target"},
	}
}

func (t *WaitForTool) Execute(args map[string]any) Result {
	kind := strings.ToLower(strings.TrimSpace(getStr(args, "type")))
	target := strings.TrimSpace(getStr(args, "target"))
	if target == "" {
		return R("error: target is required")
	}

	timeout := getFloat(args, "timeout")
	if timeout <= 0 {
		timeout = 10.0
	}
	if timeout > 30.0 {
		timeout = 30.0
	}
	interval := getFloat(args, "interval")
	if interval <= 0 {
		interval = 0.5
	}
	if interval < 0.1 {
		interval = 0.1
	}

	if kind != "window" && kind != "text" && kind != "image" {
		return R("error: type must be window, text, or image")
	}

	start := time.Now()
	for {
		found, err := waitCheck(kind, target)
		if err != nil {
			return R("error: " + err.Error())
		}
		elapsed := time.Since(start).Seconds()
		if found {
			return R(fmt.Sprintf("found after %.1fs", elapsed))
		}
		if elapsed+interval >= timeout {
			return R(fmt.Sprintf("timeout: '%s' not found after %.0fs", target, timeout))
		}
		time.Sleep(time.Duration(interval*1000) * time.Millisecond)
	}
}

func waitCheck(kind, target string) (bool, error) {
	switch kind {
	case "window":
		windows, err := enumerateWindows()
		if err != nil {
			return false, err
		}
		needle := strings.ToLower(target)
		for _, w := range windows {
			if strings.Contains(strings.ToLower(w.title), needle) {
				return true, nil
			}
		}
		return false, nil
	case "image":
		screenshotPath, err := captureRegionToFile(nil)
		if err != nil {
			return false, err
		}
		defer os.Remove(screenshotPath)
		if _, err := os.Stat(target); os.IsNotExist(err) {
			return false, fmt.Errorf("template image not found: %s", target)
		}
		shell := shellExe()
		cmd := exec.Command(shell, "-NoProfile", "-NonInteractive", "-Command", imageLocateScript,
			"-TemplatePath", target, "-ScreenshotPath", screenshotPath)
		out, _ := cmd.Output()
		result := strings.TrimSpace(string(out))
		return result != "NOTFOUND" && result != "", nil
	case "text":
		path, err := captureRegionToFile(nil)
		if err != nil {
			return false, err
		}
		defer os.Remove(path)
		text, err := runOCR(path)
		if err != nil {
			return false, err
		}
		return strings.Contains(strings.ToLower(text), strings.ToLower(target)), nil
	}
	return false, nil
}
